lib/Segments/DynamicSegments/FilterDataMapper.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace MailPoet\Segments\DynamicSegments;
+
+use MailPoet\Entities\DynamicSegmentFilterData;
+use MailPoet\Segments\DynamicSegments\Exceptions\InvalidFilterException;
+use MailPoet\Segments\DynamicSegments\Filters\WooCommerceCategory;
+use MailPoet\Segments\DynamicSegments\Filters\WooCommerceProduct;
+
+class FilterDataMapper {
+  public function map(array $data = []): DynamicSegmentFilterData {
+    switch ($this->getSegmentType($data)) {
+      case DynamicSegmentFilterData::TYPE_USER_ROLE:
+        if (empty($data['wordpressRole'])) throw new InvalidFilterException('Missing role', InvalidFilterException::MISSING_ROLE);
+        return new DynamicSegmentFilterData([
+          'segmentType' => DynamicSegmentFilterData::TYPE_USER_ROLE,
+          'wordpressRole' => $data['wordpressRole'],
+        ]);
+      case DynamicSegmentFilterData::TYPE_EMAIL:
+        return $this->createEmail($data);
+      case DynamicSegmentFilterData::TYPE_WOOCOMMERCE:
+        return $this->createWooCommerce($data);
+      default:
+        throw new InvalidFilterException('Invalid type', InvalidFilterException::INVALID_TYPE);
+    }
+  }
+
+  /**
+   * @throws InvalidFilterException
+   */
+  private function getSegmentType(array $data): string {
+    if (!isset($data['segmentType'])) {
+      throw new InvalidFilterException('Segment type is not set', InvalidFilterException::MISSING_TYPE);
+    }
+    return $data['segmentType'];
+  }
+
+  /**
+   * @throws InvalidFilterException
+   */
+  private function createEmail(array $data): DynamicSegmentFilterData {
```

Comment:
You verify the name of the action in method `createWooCommerce`, but this method can pass another than the supported action. I would prefer to check that the action value is valid. What do you think?
One more thing: If I remember right so then is `link_id` required for click and not click actions.

Reply:
It makes sense to validation for all those cases you mentioned 👍 I copied the validation from the old mapper.

Reply:
Ha! Actually, it seems that link_id is not required. Without the specific link, the segments check whether users clicked on some or didn't click at any link. So I think it will be fine to add just supported actions validation.

Reply:
I've added validation of supported actions for EmailAction filter. As for the wooCommerce filter, there is an [else branch that throws an error in case the action is not known](https://github.com/mailpoet/mailpoet/pull/3365/commits/a457c725656ecfd4efebac3f3a8e6c02eb72ce07#diff-4b71db0f185b4e8e279f557aa4723d13d273b56640c750f2cf44c074bc63eb33R71-R73).

Reply:
I'm sorry. I thought that the link is required for those actions.